File: frontend_v2.0/src/types/pipeline.ts
/** Pipeline graph types — aligned with FastAPI backend */

export type NodeKind =
  | 'document_loader'
  | 'chunker'
  | 'embedder'
  | 'indexer'
  | 'vector_store'
  | 'query'
  | 'retriever'
  | 'reranker'
  | 'prompt_augment'
  | 'llm'
  | 'output';

export interface PipelineNode {
  id: string;
  kind: NodeKind;
  label: string;
  x: number;
  y: number;
  config: Record<string, unknown>;
}

export interface PipelineEdge {
  id: string;
  source_id: string;
  source_port: string;
  target_id: string;
  target_port: string;
}

export interface PipelineGraph {
  nodes: PipelineNode[];
  edges: PipelineEdge[];
}

export interface NodeExecutionResult {
  node_id: string;
  kind: NodeKind;
  started_at_ms: number;
  finished_at_ms: number;
  latency_ms: number;
  output_summary: Record<string, unknown>;
}

export interface GraphExecutionResult {
  execution_order: string[];
  node_results: NodeExecutionResult[];
}

export interface ExecuteNodeResponse {
  kind: NodeKind;
  latency_ms: number;
  output_summary: Record<string, unknown>;
}

export interface PreviewRetrievalResponse {
  chunks: Array<Record<string, unknown>>;
  query_used: string;
  top_k: number;
  error: string | null;
}

export interface DocumentSummary {
  document_id: string;
  filename: string;
  file_type: string;
  total_chunks: number;
  total_characters: number;
  created_at?: string;
}

export interface DocumentListResponse {
  documents: DocumentSummary[];
  total: number;
}

export interface Chunk {
  chunk_id: string;
  document_id: string;
  text: string;
  index: number;
  metadata: Record<string, unknown>;
}

export interface DocumentMetadata {
  filename: string;
  file_type: string;
  total_chunks: number;
  total_characters: number;
  chunk_config: Record<string, unknown>;
}

export interface ProcessedDocument {
  document_id: string;
  metadata: DocumentMetadata;
  chunks: Chunk[];
  raw_text?: string;
}

/** 3D embedding visualization (PCA projection) — aligns with backend `models/query.py`. */
export interface Point3D {
  x: number;
  y: number;
  z: number;
  label: string;
  chunk_id: string;
  is_query: boolean;
  is_neighbor: boolean;
  score: number;
  document_id: string;
  text_preview: string;
}

export interface VisualizationData {
  points: Point3D[];
  query_point: Point3D | null;
  neighbor_indices: number[];
  edges: number[][];
  rag_mode: string;
  caption: string;
}

export interface RetrievedChunk {
  chunk_id: string;
  document_id: string;
  text: string;
  score: number;
  rank: number;
  metadata: Record<string, unknown>;
}

export interface StageMetrics {
  stage_name: string;
  latency_ms: number;
  details: Record<string, unknown>;
}

export interface QueryRequest {
  query: string;
  top_k?: number;
  index_type?: string;
  override_config?: Partial<PipelineConfig>;
  selected_documents?: string[];
  visualization_only?: boolean;
}

export interface QueryResponse {
  query: string;
  answer: string;
  retrieved_chunks: RetrievedChunk[];
  visualization: VisualizationData;
  stage_metrics: StageMetrics[];
  total_latency_ms: number;
  pipeline_config: Record<string, unknown>;
}

export interface PipelineConfig {
  rag_type: string;
  index_type: string;
  retrieval_strategy: string;
  top_k: number;
  hnsw_m: number;
  hnsw_ef_construction: number;
  hnsw_ef_search: number;
  ivf_nlist?: number;
  pq_m?: number;
  pq_nbits?: number;
  enable_reranking: boolean;
  rerank_top_n: number;
  enable_agentic: boolean;
  temperature: number;
  max_tokens: number;
  openrouter_api_key?: string;
  llm_model?: string;
  embedding_model?: string;
  embedding_dim?: number;
}

File: backend/models/pipeline.py
"""Pipeline configuration Pydantic models."""
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class IndexType(str, Enum):
    FLAT = "flat"
    HNSW = "hnsw"
    IVF = "ivf"
    PQ = "pq"


class RetrievalStrategy(str, Enum):
    VECTOR = "vector"
    GRAPH = "graph"


class RagType(str, Enum):
    NAIVE = "naive"
    ADVANCED = "advanced"
    ITERATIVE = "iterative"
    AGENTIC = "agentic"
    HYBRID = "hybrid"
    VECTORLESS = "vectorless"
    GRAPH = "graph"


class PipelineConfig(BaseModel):
    rag_type: RagType = RagType.NAIVE
    index_type: IndexType = IndexType.FLAT
    retrieval_strategy: RetrievalStrategy = RetrievalStrategy.VECTOR
    top_k: int = Field(default=5, ge=1, le=50)
    # HNSW params
    hnsw_m: int = Field(default=32, ge=4, le=128)
    hnsw_ef_construction: int = Field(default=200, ge=50, le=1000)
    hnsw_ef_search: int = Field(default=64, ge=10, le=500)
    # IVF params
    ivf_nlist: int = Field(default=10, ge=1, le=100)
    # PQ params
    pq_m: int = Field(default=8, ge=2, le=64)
    pq_nbits: int = Field(default=8, ge=4, le=8)
    # Reranking (Phase 2)
    enable_reranking: bool = False
    rerank_top_n: int = Field(default=10, ge=1, le=50)
    enable_agentic: bool = False
    # Generation
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: int = Field(default=512, ge=50, le=4096)
    openrouter_api_key: Optional[str] = Field(default=None, description="OpenRouter API Key")
    llm_model: str = Field(default="openai/gpt-4o-mini", description="OpenRouter Model ID")
    embedding_model: str = Field(default="openai/text-embedding-3-small", description="Embedding Model ID")
    embedding_dim: int = Field(default=1536, description="Embedding Dimension")


class StageMetrics(BaseModel):
    stage_name: str
    latency_ms: float
    details: dict = Field(default_factory=dict)


class PipelineTypesResponse(BaseModel):
    index_types: list[dict]
    retrieval_strategies: list[dict]


class NodeKind(str, Enum):
    """Canonical node types for the visual RAG pipeline builder."""

    DOCUMENT_LOADER = "document_loader"
    CHUNKER = "chunker"
    EMBEDDER = "embedder"
    INDEXER = "indexer"
    QUERY = "query"
    RETRIEVER = "retriever"
    VECTOR_STORE = "vector_store"
    RERANKER = "reranker"
    PROMPT_AUGMENT = "prompt_augment"
    LLM = "llm"
    OUTPUT = "output"


class NodePortDirection(str, Enum):
    INPUT = "input"
    OUTPUT = "output"


class PipelineNode(BaseModel):
    """Single node instance on the visual canvas."""

    id: str
    kind: NodeKind
    label: str
    x: float = 0.0
    y: float = 0.0
    config: dict = Field(default_factory=dict)


class PipelineEdge(BaseModel):
    """Directed edge between two node ports."""

    id: str
    source_id: str
    source_port: str = "out"
    target_id: str
    target_port: str = "in"


class PipelineGraph(BaseModel):
    """Full DAG description sent from the frontend."""

    nodes: list[PipelineNode]
    edges: list[PipelineEdge]


class NodeExecutionResult(BaseModel):
    """Per-node execution metadata returned to the frontend."""

    node_id: str
    kind: NodeKind
    started_at_ms: float
    finished_at_ms: float
    latency_ms: float
    output_summary: dict = Field(default_factory=dict)


class GraphExecutionResult(BaseModel):
    """Execution summary for a single graph run."""

    execution_order: list[str]
    node_results: list[NodeExecutionResult]


class ExecuteNodeRequest(BaseModel):
    """Run a single canvas node in isolation (optional explicit inputs for experiments)."""

    node: PipelineNode
    inputs: dict = Field(default_factory=dict)


class ExecuteNodeResponse(BaseModel):
    """Result of execute-node (same shape as one NodeExecutionResult without node_id/timestamps)."""

    kind: NodeKind
    latency_ms: float
    output_summary: dict = Field(default_factory=dict)


class VerifyOpenRouterKeyRequest(BaseModel):
    """Validate an OpenRouter API key (not persisted by this request)."""

    api_key: str = Field(..., min_length=1, max_length=4000)


class CredentialsUpdate(BaseModel):
    """Update API credentials in memory without triggering index rebuild (use /configure for full apply + rebuild)."""

    openrouter_api_key: Optional[str] = None


class PreviewRetrievalRequest(BaseModel):
    query: str = Field(..., min_length=1, max_length=4000)
    top_k: int = Field(default=5, ge=1, le=50)


class PreviewRetrievalResponse(BaseModel):
    chunks: list[dict] = Field(default_factory=list)
    query_used: str = ""
    top_k: int = 5
    error: str | None = None
